### ISSUE TITLE
Implementation Request : Regularizing RNNs by Stabilizing Activations

### ISSUE THREAD

**Issue by nicholas-leonard** [Element-Research/rnn#88]
It would be nice if we could implement this for rnn package : http://arxiv.org/abs/1511.08400

**Comment by juesato** [comment]
How would you implement this? I could modify the `_accGradParameters` in Recurrent.lua directly, and with a little more work get the gradients for each of the gates and update `LSTM.lua`, but these seems like a bit of a hack. 

Are there any other examples within the library of how you've implemented regularization? Dropout makes sense as a separate layer, but I don't think this does.

**Comment by nicholas-leonard** [comment]
@juesato I just spoke to the author of the paper. The most general way would be to implement it as its own `nn.Stabilizer` layer. It would inherit AbstractRecurrent. When you build your model you would do : 
```
rm = nn.Sequential()
   :add(nn.LSTM())
   :add(nn.Stabilizer())
   :add(nn.LSTM())
model = nn.Recursor(rm) or nn.Sequencer(rm)
```

**Comment by juesato** [comment]
@nicholas-leonard I'm confused how this works as a layer. Wouldn't this regularization technique operate on the weights of the layer itself (change the weight update equations), as opposed to operating on the output activations of the layer (as `Dropout` layers do)?

**Comment by nicholas-leonard** [comment]
@juesato So basically, the `NormStabilizer` would look like this (see `CopyGrad` implementation below):

```lua
local NS, parent = torch.class("nn.NormStabilizer", "nn.AbstractRecurrent")

-- setting the recurrentModule to CopyGrad makes it save the output (h[t]) 
-- and copy gradOutput for each step.
function NS:__init(beta)
   ...
   self.recurrentModule = nn.CopyGrad()
   ...
end

-- the usual AbstractRecurrent subclass code. 
function NS:updateOutput(input)
   local output
   if self.train ~= false then
      self:recycle()
      local recurrentModule = self:getStepModule(self.step)
      -- the actual forward propagation
      output = recurrentModule:updateOutput(input)
   else
      output = self.recurrentModule:updateOutput(input)
   end
   
   self.outputs[self.step] = output
   
   self.output = output
   
   self.step = self.step + 1
   self.gradPrevOutput = nil
   self.updateGradInputStep = nil
   self.accGradParametersStep = nil
   
   return self.output
end

function NS:_updateGradInput(input, gradOutput)
    -- this is where you implement the norm-stabilizer cost function.
    -- first you need h[t] and h[t-1] :
    local hiddenModule = self:getStepModule(self.updateGradInputStep)
    local prevhiddenModule = self:getStepModule(self.updateGradInputStep)
   local hiddenstate = hiddenModule.output
   local prevhiddenstate = prevhiddenModule.output
   -- then when you implement the norm-stabilizer cost function you will end up with two gradients 
   -- which you directly add to the respective CopyGrad.gradInput tensors
   local hiddenModule.gradInput:add( ...d norm stabilizer / d h[t]...)
   local prevHiddenModule.gradInput:add( ...d norm stabilizer / d h[t-t]...). 
   return hiddenModule.gradInput
end

...
```

where `nn.CopyGrad` looks like this : 

```lua
local CopyGrad, _ = torch.class('nn.CopyGrad', 'nn.Identity')

function CopyGrad:updateGradInput(input, gradOutput)
   self.gradInput:resizeAs(gradOutput):copy(gradOutput)
   return self.gradInput
end
```

**Comment by juesato** [comment]
@nicholas-leonard Got it, working on this now.

**Comment by juesato** [comment]
@nicholas-leonard I'm working on this. Do you have suggestions for how to test this? I've been modifying the `simple-sequencer-network.lua` network, but it just converges slower (which I expect, but also doesn't tell me it's working).

Also, I think https://github.com/Element-Research/rnn/blob/master/AbstractRecurrent.lua#L235, there's a bug, since I'm pretty sure neither `nn.Container` nor `nn.Module` define a `__tostring__()` method. It's not really important, since it never gets called, but I figure the else case should just be something like `return "nn.AbstractRecurrent"` instead.

**Comment by nicholas-leonard** [comment]
For testing, what I like to do is build a unit test where I compare the forward/backward of the module with equivalent code written differently. This could either be code written with a bunch of modules (e.g. https://github.com/Element-Research/dpnn/blob/master/test/test.lua#L715-L732)  or a just using pure torch/Lua code (e.g. https://github.com/Element-Research/dpnn/blob/master/test/test.lua#L463-L477).

Basically, your `input` is a sequence like `{torch.randn(3,4), torch.randn(3,4),...,}` and so is your `gradOutput`. Your module will give you some gradInput which your own boiler-plate test code should match. The idea behind it is that it is difficult to make the same exact mistake twice. So when both your boiler-plate code and your main module have the same behaviour, it probably means that they are both right.


Also, not a bug :
```lua
th> tostring(nn.Module())
nn.Module	
th> tostring(nn.Container())
nn.Container	
```

**Comment by nicholas-leonard** [comment]
@juesato How is this moving along? Don't hesitate to submit a Pull Request with anything you got.

**Comment by juesato** [comment]
Thanks for the poke and sorry for the delay. I need to write tests and clean up code, and will update tomorrow.

**Comment by juesato** [comment]
@nicholas-leonard 

Just to update, right now this is what I have for `NormStabilizer.lua`. I'm not convinced this is right since it takes longer to converge on the `simple-sequencer-network.lua` example. I'm writing a different implementation as a `Decorator` to wrap around an `AbstractRecurrent` instance, and a numerical gradient check.

```
function NS:__init(beta, rho)
   parent.__init(self, rho or 9999)
   self.recurrentModule = nn.CopyGrad()
   self.beta = beta
end

function NS:_accGradParameters(input, gradOutput, scale)
   -- No parameters to update
end

function NS:updateOutput(input)
   local output
   if self.train ~= false then
      self:recycle()
      local recurrentModule = self:getStepModule(self.step)
      output = recurrentModule:updateOutput(input)
   else
      output = self.recurrentModule:updateOutput(input)
   end

   self.outputs[self.step] = output

   self.output = output
   self.step = self.step + 1
   self.gradPrevOutput = nil
   self.updateGradInputStep = nil
   self.accGradParametersStep = nil

   return self.output
end

function NS:_updateGradInput(input, gradOutput)
   -- First grab h[t] and h[t+1] :
   -- backward propagate through this step
   local gradInput = self.recurrentModule:updateGradInput(input, gradOutput)

   local hiddenModule = self:getStepModule(self.updateGradInputStep)
   hiddenModule.gradInput = gradInput
   if self.updateGradInputStep + 1 <= self.step - 1 then
      local prevHiddenModule = self:getStepModule(self.updateGradInputStep - 1)
      local nextHiddenModule = self:getStepModule(self.updateGradInputStep + 1)
      local hiddenState = hiddenModule.output
      local prevHiddenState = prevHiddenModule.output
      local nextHiddenState = nextHiddenModule.output
      -- Add norm stabilizer cost function directly to respective CopyGrad.gradInput tensors
      local dRegdHidden =  self.beta * 2 * (hiddenState:norm()-prevHiddenState:norm())
      hiddenModule.gradInput:add( torch.mul(hiddenState, dRegdHidden / hiddenState:norm()) )
      local dReg2dHidden = self.beta * 2 * (nextHiddenState:norm() - hiddenState:norm())
      hiddenModule.gradInput:add( torch.mul(hiddenState, dReg2dHidden / hiddenState:norm()) )
   end
   return hiddenModule.gradInput
end

function NS:__tostring__()
   return "nn.NormStabilizer"
end
```

**Comment by juesato** [comment]
Update: the code above is definitely wrong. Updating the weights according to the numerically approximated gradient converges pretty fast.

**Comment by juesato** [comment]
@nicholas-leonard I can test this in the case where the outputs of the RNN feed directly into a `Criterion` by using a modified version of `SequencerCriterion` which implements the regularization function.

The code for SequencerCriterion with norm stabilization is below:

```lua
------------------------------------------------------------------------
--[[ SequencerCriterionReg ]]--
-- Modified version of SequencerCriterion that adds norm stabilization with beta = 5.0
------------------------------------------------------------------------
local SequencerCriterion, parent = torch.class('nn.SequencerCriterionReg', 'nn.Criterion')

function SequencerCriterion:__init(criterion)
   parent.__init(self)
   self.criterion = criterion
   if torch.isTypeOf(criterion, 'nn.ModuleCriterion') then
      error("SequencerCriterion shouldn't decorate a ModuleCriterion. "..
         "Instead, try the other way around : "..
         "ModuleCriterion decorates a SequencerCriterion. "..
         "Its modules can also be similarly decorated with a Sequencer.")
   end
   self.clones = {}
   self.gradInput = {}
end

function SequencerCriterion:getStepCriterion(step)
   assert(step, "expecting step at arg 1")
   local criterion = self.clones[step]
   if not criterion then
      criterion = self.criterion:clone()
      self.clones[step] = criterion
   end
   return criterion
end

function SequencerCriterion:updateOutput(inputTable, targetTable)
   self.output = 0
   for i,input in ipairs(inputTable) do
      local criterion = self:getStepCriterion(i)
      self.output = self.output + criterion:forward(input, targetTable[i])
      if i > 1 then
         local reg = 0
         for j=1,input:size(1) do
            reg = reg + ((input[j]:norm() - inputTable[i-1][j]:norm())^2)
         end
         self.output = self.output + 5.0 * reg / input:size(1)
      end
   end
   
   return self.output
end

function SequencerCriterion:updateGradInput(inputTable, targetTable)
   self.gradInput = {}
   
   for i,input in ipairs(inputTable) do
      local criterion = self:getStepCriterion(i)
      self.gradInput[i] = criterion:backward(input, targetTable[i])

      if i > 1 then
         for j=1,input:size(1) do
            self.gradInput[i]:add(torch.mul(inputTable[i][j], 5.0 * 2 * (inputTable[i][j]:norm() - inputTable[i-1][j]:norm()) / inputTable[i][j]:norm()) )
         end
      end
      if i < #inputTable then
         for j=1,input:size(1) do
            self.gradInput[i]:add(torch.mul(inputTable[i][j], 5.0 * -2 * (inputTable[i+1][j]:norm() - inputTable[i][j]:norm()) / inputTable[i][j]:norm()) )
         end
      end
   end
   
   return self.gradInput
end
```

And I've been using this script to test it. It compares the numerical gradient to the analytically computed gradient, and the error is in the 1e-5 to 1e-6 range.

```lua
------------------------------------------------------------------------
--[[ SequencerCriterionReg ]]--
-- Modified version of SequencerCriterion that adds norm stabilization with beta = 5.0
------------------------------------------------------------------------
local SequencerCriterion, parent = torch.class('nn.SequencerCriterionReg', 'nn.Criterion')

function SequencerCriterion:__init(criterion)
   parent.__init(self)
   self.criterion = criterion
   if torch.isTypeOf(criterion, 'nn.ModuleCriterion') then
      error("SequencerCriterion shouldn't decorate a ModuleCriterion. "..
         "Instead, try the other way around : "..
         "ModuleCriterion decorates a SequencerCriterion. "..
         "Its modules can also be similarly decorated with a Sequencer.")
   end
   self.clones = {}
   self.gradInput = {}
end

function SequencerCriterion:getStepCriterion(step)
   assert(step, "expecting step at arg 1")
   local criterion = self.clones[step]
   if not criterion then
      criterion = self.criterion:clone()
      self.clones[step] = criterion
   end
   return criterion
end

function SequencerCriterion:updateOutput(inputTable, targetTable)
   self.output = 0
   for i,input in ipairs(inputTable) do
      local criterion = self:getStepCriterion(i)
      self.output = self.output + criterion:forward(input, targetTable[i])
      if i > 1 then
         local reg = 0
         for j=1,input:size(1) do
            reg = reg + ((input[j]:norm() - inputTable[i-1][j]:norm())^2)
         end
         self.output = self.output + 5.0 * reg / input:size(1)
      end
   end
   
   return self.output
end

function SequencerCriterion:updateGradInput(inputTable, targetTable)
   self.gradInput = {}
   
   for i,input in ipairs(inputTable) do
      local criterion = self:getStepCriterion(i)
      self.gradInput[i] = criterion:backward(input, targetTable[i])

      if i > 1 then
         for j=1,input:size(1) do
            self.gradInput[i]:add(torch.mul(inputTable[i][j], 5.0 * 2 * (inputTable[i][j]:norm() - inputTable[i-1][j]:norm()) / inputTable[i][j]:norm()) )
         end
      end
      if i < #inputTable then
         for j=1,input:size(1) do
            self.gradInput[i]:add(torch.mul(inputTable[i][j], 5.0 * -2 * (inputTable[i+1][j]:norm() - inputTable[i][j]:norm()) / inputTable[i][j]:norm()) )
         end
      end
   end
   
   return self.gradInput
end
```

I'm not sure why this same type of thing isn't working when I implement this in the `NormStabilizer.lua` layer

**Comment by nicholas-leonard** [comment]
@juesato Thanks again for taking the time to implement this!
